src/simplewallet/simplewallet.cpp

```diff
@@ -1946,9 +1952,109 @@ bool simple_wallet::get_address_from_str(const std::string &str, cryptonote::acc
+      std::vector<uint64_t> absolute_offsets = cryptonote::relative_output_offsets_to_absolute(in_key.key_offsets);
+      // get block heights from which those ring member keys originated
+      COMMAND_RPC_GET_OUTPUTS_BIN::request req = AUTO_VAL_INIT(req);
+      req.outputs.resize(absolute_offsets.size());
+      for (size_t j = 0; j < absolute_offsets.size(); ++j)
+      {
+        req.outputs[j].amount = in_key.amount;
+        req.outputs[j].index = absolute_offsets[j];
+      }
+      COMMAND_RPC_GET_OUTPUTS_BIN::response res = AUTO_VAL_INIT(res);
+      bool r = net_utils::invoke_http_bin_remote_command2(m_wallet->get_daemon_address() + "/get_outs.bin", req, res, m_http_client);
+      err = interpret_rpc_response(r, res.status);
+      if (!err.empty())
+      {
+        fail_msg_writer() << tr("failed to get output: ") << err;
+        return false;
+      }
+      // make sure that returned block heights are less than blockchain height
+      for (auto& res_out : res.outs)
+      {
+        if (res_out.height >= blockchain_height)
+        {
+          fail_msg_writer() << tr("output key's originating block height shouldn't be higher than the blockchain height");
+          return false;
+        }
+      }
+      ostr << tr("\nOriginating block heights: ");
+      for (size_t j = 0; j < absolute_offsets.size(); ++j)
+        ostr << tr(j == source.real_output ? " *" : " ") << res.outs[j].height;
+      spent_key_height[i] = res.outs[source.real_output].height;
+      spent_key_txid  [i] = res.outs[source.real_output].txid;
+      // visualize the distribution, using the code by moneroexamples onion-monero-viewer
+      const uint64_t resolution = 79;
+      std::string ring_str(resolution, '_');
+      for (size_t j = 0; j < absolute_offsets.size(); ++j)
+      {
+        uint64_t pos = (res.outs[j].height * resolution) / blockchain_height;
+        ring_str[pos] = 'o';
+      }
+      uint64_t pos = (res.outs[source.real_output].height * resolution) / blockchain_height;
+      ring_str[pos] = '*';
+      ostr << tr("\n|") << ring_str << tr("|\n");
+    }
+    // warn if rings contain keys originating from the same tx or temporally very close block heights
+    bool are_keys_from_same_tx      = false;
+    bool are_keys_from_close_height = false;
+    for (size_t i = 0; i < tx.vin.size(); ++i) {
+      for (size_t j = i + 1; j < tx.vin.size(); ++j)
+      {
+        if (spent_key_txid[i] == spent_key_txid[j])
+          are_keys_from_same_tx = true;
+        if (std::abs(spent_key_height[i] - spent_key_height[j]) < 5)
+          are_keys_from_close_height = true;
+      }
+    }
+    if (are_keys_from_same_tx || are_keys_from_close_height)
+    {
+      ostr
+        << tr("\nWarning: Some input keys being spent are from ")
+        << tr(are_keys_from_same_tx ? "the same transaction" : "blocks that are temporally very close")
+        << tr(", which can break the anonymity of ring signature. Make sure this is intentional!");
```

Comment:
This is accurate. I just know this will result in loads of scared people coming to ask the same thing again and again though, because they don't understand.